cocos/audio/AudioEngine.cpp

```diff
@@ -67,101 +70,98 @@ AudioEngine::AudioEngineThreadPool* AudioEngine::s_threadPool = nullptr;
 class AudioEngine::AudioEngineThreadPool
 {
 public:
-    AudioEngineThreadPool(bool detach)
-        : _numThread(6)
-        , _detach(detach)
+    AudioEngineThreadPool(bool detach, int threads = 4)
+        : _detach(detach)
+        , _stop(false)
     {
-        s_threadPool = this;
-
-        _threads.reserve(_numThread);
-        _tasks.reserve(_numThread);
+        _scheduler = Director::getInstance()->getScheduler();
 
-        for (int index = 0; index < _numThread; ++index) {
-            _tasks.push_back(nullptr);
-            _threads.push_back(std::thread(std::bind(&AudioEngineThreadPool::threadFunc, this, index)));
+        for (int index = 0; index < threads; ++index)
+        {
+            _workers.push_back(std::thread(std::bind(&AudioEngineThreadPool::threadFunc, this)));
             if (_detach)
             {
-                _threads[index].detach();
+                _workers[index].detach();
             }
         }
     }
 
-    void addTask(const std::function<void()> &task){
-        _taskMutex.lock();
-        int targetIndex = -1;
-        for (int index = 0; index < _numThread; ++index) {
-            if (_tasks[index] == nullptr) {
-                targetIndex = index;
-                _tasks[index] = task;
-                break;
-            }
-        }
-        if (targetIndex == -1) {
-            _tasks.push_back(task);
-            _threads.push_back(std::thread(std::bind(&AudioEngineThreadPool::threadFunc, this, _numThread)));
-            if (_detach)
-            {
-                _threads[_numThread].detach();
-            }
-            _numThread++;
-        }
-        _taskMutex.unlock();
-        _sleepCondition.notify_all();
+    void addTask(const std::function<bool()> &task, AudioPreloadCallback callback){
+        std::unique_lock<std::mutex> lk(_queueMutex);
+        _taskQueue.push(task);
```

Comment:
please use `emplace` instead of `push_back`: http://www.cplusplus.com/reference/queue/queue/emplace/